pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -522,20 +535,26 @@ func getNodeNameFromHostname(nodes *v1.NodeList, hostname string) (string, bool)
 }
 
 func (c *Cluster) startMon(m *monConfig, hostname string) error {
-	rs := c.makeReplicaSet(m, hostname)
-	logger.Debugf("Starting mon: %+v", rs.Name)
-	_, err := c.context.Clientset.Extensions().ReplicaSets(c.Namespace).Create(rs)
+	// If we determine the legacy replicaset exists, delete it so we can start the new deployment in its place
+	if err := k8sutil.DeleteReplicaSet(c.context.Clientset, c.Namespace, m.ResourceName); err != nil {
```

Comment:
Is "legacy replicaset" referring to the `rook-ceph-mon0` (and so on) ReplicaSets?
If so, why does this need to be checked every time?

As far as I see, the old mons `rook-ceph-mon1` (and so on) are replaced with e.g. `rook-ceph-mon-a` Deployments. The new mon Deployments are using a data directory which uses their new name, right?
Because of that I think it should be added that the data directories of the old mons should be deleted after all mons have been upgraded and the cluster has been verified to work.

I think we should look into moving such "migrations" to a separate file so that they are not run every time this action is triggered, but only "once" during operator start.
What do you think? (This would be out of scope for this PR).

Reply:
Mons that already exist will actually retain the same name. So if we had the `rook-ceph-mon1` replicaset, the operator will delete that replicaset and start a deployment named `rook-ceph-mon1` and the data dir will also have the same name. Only when a new mon is created/failed over will the new naming style of `rook-ceph-mon-d` be used.

`startMon()` should only be called once at startup and then during mon failover, right? While deletion of the replicaset only needs to be performed once ever, it is simpler to just try it every time than to keep a record of i being done.